src/python/dxpy/app_builder.py

```diff
@@ -212,58 +161,199 @@ def upload_resources(src_dir, project=None, folder='/', ensure_upload=False):
-        # contents, and reuse it if possible. The resource bundle
-        # carries a property 'resource_bundle_checksum' that indicates
-        # the checksum; the way in which the checksum is computed is
-        # given in the documentation of _directory_checksum.
+        # Copy the code previously in _directory_checksum here and we will create
+        # the tar file while performing the checksum.  If the checksum matches 
+        # (and ensure_upload is False), then we will use the existing file,
+        # otherwise, we will compress and upload the tarball.
+        
+        with tempfile.NamedTemporaryFile(suffix=".tar") as tar_fh:
+        
+            output_sha1 = hashlib.sha1()
+        
+            for dirname, subdirs, files in os.walk(resources_dir):
+                if not dirname.startswith(resources_dir):
+                    raise AssertionError('Expected %r to start with root directory %r' % (dirname, resources_dir))
+
+                # Add an entry for the directory itself
+                relative_dirname = dirname[len(resources_dir):]
+                dir_stat = os.lstat(dirname)
+                if not relative_dirname.startswith('/'):
+                    relative_dirname = '/' + relative_dirname
+
+                fields = [relative_dirname, str(dir_stat.st_mode), str(int(dir_stat.st_mtime * 1000))]
+                output_sha1.update(b''.join(s.encode('utf-8') + b'\0' for s in fields))
+                
+                # add an entry in the tar file for the current directory, but
+                # do not recurse!
+                subprocess.check_call(['tar', '-C', resources_dir, '-rf', tar_fh.name, '--no-recursion', '.' + relative_dirname])
+                
+                # Canonicalize the order of subdirectories; this is the order in
+                # which they will be visited by os.walk
+                subdirs.sort()
+                
+                # check the subdirectories for symlinks.  We should throw an error
+                # if there are any links that point outside of the directory (unless
+                # --force-symlinks is given).  If a link is pointing internal to 
+                # the directory (or --force-symlinks is given), we should add it 
+                # as a file.
+                for subdir_name in subdirs:
+                    dir_path = os.path.join(dirname, subdir_name)
+
+                    # If we do have a symlink,                     
+                    if os.path.islink(dir_path):
+                        # Let's get the pointed-to path to ensure that it is
+                        # still in the directory
+                        link_target = os.readlink(dir_path)
+                        is_local=(not os.path.isabs(link_target))
+                        
+                        dir_realpath=os.path.abspath(os.path.join(dirname, link_target))
+                        # make sure that the path NEVER extends outside the resources directory!
```

Comment:
Correct; the idea is to transfer links that would break; this would fall under case 8.  I agree that it's a pathological example to have a link like that, but nothing stops someone from doing something silly like that.

I agree that it's a lot of trouble to protect people from this pathological use case, but I don't think there's a huge computational burden.  It would be my preference to fail early, when possible, rather than failing when trying to reference the link.  Should we put some debugging statements for when links are dereferenced (and if so, what's the "DNAnexus way" of doing so)?